### ISSUE TITLE
Better understand of apollo-client read and merge functions to get data from the cache when possible

### ISSUE THREAD

**Issue by decosvaldo** [apollographql/apollo-client#6990]
Hi everybody. I'm stuck on how to get data from the apollo cache when possible but also query from the GraphQL server when necessary.

I'm using an implementation of prisma/nexus GraphQL server, so my pagination is performed by using limit and first/last parameters for pagination. I also have the possibility to use where and orderby.

My project is a website where we can perform album reviews, rating, etc. 

I have a main component of the application where I need to display the last 15 albums reviews without defining a specific album and also with a button attached to the fetchMore function. So this is my query:

```
query($first:Int,$skip:Int)
    {
        albumreviews(first:$first,skip:$skip,orderBy:{
            createdAt: desc
        },where:{
            status:{
                equals: 1
            }
        }){
            id
            title
            createdAt
            createdBy {
                id
                name
            }
            album {
                id
                name
                year
                artist {
                    id
                    name
                    genres {
                        id
                        name
                    }
                }
                votes{
                    id
                    vote
                    createdBy {
                        id
                    }
                }
                blobCover
            }
        }
    }


const { loading, error, data, fetchMore } = useQuery(REVIEWS, {
        variables: {
            first: 10,
            skip: 0
        }
    })

```
To make the apollo 3 client work with first and skip args as limit and offset, I've changed the default offset pagination function and created this one:

```
export function myOffsetLimitPagination<T = Reference>(
    keyArgs: KeyArgs = false,
    ): FieldPolicy<T[]> {
    return {
        keyArgs,
        merge(existing, incoming, { args }) {
            if(args.skip){
                const merged = existing ? existing.slice(0) : [];
                const start = args ? args.skip : merged.length;
                const end = start + incoming.length;
                for (let i = start; i < end; ++i) {
                    merged[i] = incoming[i - start];
                }
                return merged
            }else{
                return incoming
            }
        }
    };
}
```

I also have an album component, where I can see more info about an album and all of it's reviews. So I will perform this query:

```
query($id:Int)
    {
        albumreviews(orderBy:{
            createdAt: desc
        },where:{
            AND:[{
                status:{
                    equals: 1
                },
                album: {
                    equals: $id
                }
            }]
        }){
            id
            title
            createdAt
            createdBy {
                id
                name
            }
            album {
                id
                name
                year
                artist {
                    id
                    name
                    genres {
                        id
                        name
                    }
                }
                votes{
                    id
                    vote
                    createdBy {
                        id
                    }
                }
                blobCover
            }
        }
    }


const { loading, error, data } = useQuery(REVIEWSBYALBUM, {
        variables: {
            id: props.id
        }
    })

```

With this in thought, if I access the main component first, I can see the first albums reviews and I also receive more when I click the fetchMore button. Yes! 
But, if I go to the album component, I will not see the reviews from that album, I will see the query that was executed on the main component. And the same happens if I access the album component first: the main component will only present the reviews from the album that was on the first query.

So, this is what I need to understand and learn how to fix:

How can I get and diplay only the correct information from the cache (if is in there) and also get the remaining information from the network (also if necessary) and display the correct and necessary information on each component? I know that I have the option to force the cache to also use the network, but I believe that something is missing on my code since the album component it is diplaying information that it shouldn't, since there is a "where" filter there. And vice-versa since the main component also presents wrong information with a query with the same __typename happens elsewhere. And I also want to use the best option for user experience and performance.

I'm using the default fetchPolicy from the apollo-client 3.

**Comment by jacdx** [comment]
Having this same issue. Docs are not helpful, seems to be a lot of confusion around 3.0's pagination techniques, other than for the simplest use cases. Did you figure this out?

**Comment by decosvaldo** [comment]
Nope. Had to put my project on hold because of this. Can't find information on how to move forward on ever understand how to improve my code.

**Comment by mgan59** [comment]
I'm crawling all over the docs, spectrum, and googling, etc. Trying to understand how to correctly use the `cache`.  I have to say the documentation though vast and detailed is great reference documentation.  I understand the fundamental approach of what `cache` solves and how it can be utilized. However, it is over-whelming and not actually useable for someone looking to get started with setting up cache. I've started small just want to read/write some data to the cache and even that is proving challenging.  I'm sure I'll get it working.  But felt it worth piling on in the issues here that there is friction on `cache` setup-usage.

**Comment by hwillson** [comment]
The docs have been updated quite a bit since this issue was opened. Let us know if you still think this is an issue. Thanks!
